components/add_user_to_group_multiselect/add_user_to_group_multiselect.tsx

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+
+import {Client4} from 'mattermost-redux/client';
+import {RelationOneToOne} from 'mattermost-redux/types/utilities';
+import {ActionResult} from 'mattermost-redux/types/actions';
+import {UserProfile} from 'mattermost-redux/types/users';
+
+import {filterProfilesStartingWithTerm, isGuest} from 'mattermost-redux/utils/user_utils';
+import {displayEntireNameForUser, localizeMessage} from 'utils/utils.jsx';
+import ProfilePicture from 'components/profile_picture';
+import MultiSelect, {Value} from 'components/multiselect/multiselect';
+import AddIcon from 'components/widgets/icons/fa_add_icon';
+import GuestBadge from 'components/widgets/badges/guest_badge';
+import BotBadge from 'components/widgets/badges/bot_badge';
+
+import Constants from 'utils/constants';
+
+const USERS_PER_PAGE = 50;
+const MAX_SELECTABLE_VALUES = 20;
+
+type UserProfileValue = Value & UserProfile;
+
+export type Props = {
+    multilSelectKey: string;
+    userStatuses: RelationOneToOne<UserProfile, string>;
+    focusOnLoad?: boolean;
+
+    // Used if we are adding new members to an existing group
+    groupId?: string;
+
+    // onSubmitCallback takes an array of UserProfiles and should set usersToAdd in state of parent component
+    onSubmitCallback: (userProfiles?: UserProfile[]) => Promise<void>;
+    addUserCallback?: (userProfiles: UserProfile[]) => void;
+    deleteUserCallback?: (userProfiles: UserProfile[]) => void;
+
+    // These are the optinoal search parameters
+    searchOptions?: any;
+
+    // Dictionaries of userid mapped users to exclude or include from this list
+    excludeUsers?: Record<string, UserProfileValue>;
+    includeUsers?: Record<string, UserProfileValue>;
+
+    profiles: UserProfileValue[];
+
+    savingEnabled: boolean;
+    buttonSubmitText?: string;
+    buttonSubmitLoadingText?: string;
+    backButtonClick?: () => void;
+    backButtonClass?: string;
+
+    actions: {
+        getProfiles: (page?: number, perPage?: number) => Promise<ActionResult>;
+        getProfilesNotInGroup: (groupId: string, page?: number, perPage?: number) => Promise<ActionResult>;
+        loadStatusesForProfilesList: (users: UserProfile[]) => void;
+        searchProfiles: (term: string, options: any) => Promise<ActionResult>;
+    };
+}
+
+type State = {
+    values: UserProfileValue[];
+    term: string;
+    saving: boolean;
+    loadingUsers: boolean;
+}
+
+export default class AddUserToGroupMultiSelect extends React.PureComponent<Props, State> {
```

Comment:
nit: This component has 270 LOC, I don't love big components, I don't know if there is any way to reduce that, maybe moving the render option in its own component. Again, this is not an stopper, it is just a perception that this component is a bit bigger than what I would like.

Reply:
Moved the render option part into its own functional component, this component is now 244 LOC. I don't think there is much else I can do to reduce it.